View pokemon sprites on box card

YaManicKill opened the pull request:
Fixes #685

TODO:

- [x] Add pokemon to box if you add it and it isn't the currently selected box
- [x] Change base to master when #671 is merged in

![c1884f21406be89870d5810605937586](https://cloud.githubusercontent.com/assets/2162657/23073201/a679f134-f52c-11e6-9756-f95e883477b4.png)

Raia commented:
I've wanted this forever \o/

I'd honestly drop the description, because when I did that mini-survey and looked around at people's profiles, not many people used them. Plus, the idea of "description" is usually something you'd see when viewing a full page. The title + sprites should be sufficient for distinguishing boxes.

not-an-aardvark commented on client/pokemon/pokemon-icon.view.html:
```diff
@@ -0,0 +1,4 @@
+<div class="icon-container" ng-init="pokemon.parseCardViewProps()" flex
+  ng-class="['spr-box', pokemon.spriteClass, {'pokemon-egg': pokemon.data.isEgg}]" >
+  <span ng-if="::pokemon.data.isEgg" class="spr-box spr-box-egg"></span>
```
Have you tested the `isEgg` logic here? I don't think the `isEgg` property will be returned from the server, because it's not in `POKEMON_FIELDS_FOR_CARD`. So the egg icon will never appear.

YaManicKill replied:
I haven't actually, that's a good point I tested how it looked (by changing the data locally) but not from an actual egg. I shall do that today.

SnowPhoenix replied:
Confirming bug and...erm, [fix...?](http://i.imgur.com/8VHM3Hg.png)

YaManicKill replied:
What am I looking at here?

SnowPhoenix replied:
I'm not sure if we want the egg in that position, i.e. covering the Pokémon's face. Hence the uncertainty as to whether we should consider that a complete fix or not.

For comparison, we usually place the egg sprite [behind the Pokémon](http://i.imgur.com/g3SCADy.png), but we might not have enough room to keep it noticeable on a grid like this.

YaManicKill replied:
Ah right. I'll see what I can do :-)

YaManicKill replied:
Aight, done both of these.

not-an-aardvark commented on client/pokemon/pokemon.ctrl.js:
```diff
@@ -31,13 +31,7 @@ module.exports = class Pokemon {
-    this.paddedEsv = this.data.esv.toString().padStart(4, '0');
-    this.parsedNickname = replace3dsUnicodeChars(this.data.nickname);
-
+  parseCardViewProps () {
```
If you're renaming `parseBoxViewProps` to `parseCardViewProps`, make sure to also change the calls to `parseBoxViewProps`, otherwise the card view will probably be broken.

SnowPhoenix replied:
See line 46. He's not renaming parseBoxViewProps. He's just separating a portion of it so it can be called elsewhere (client/pokemon/pokemon-icon.view.html).

YaManicKill replied:
Ah, you got to this before I did.

Exactly :-)

not-an-aardvark commented on client/box/box.ctrl.js:
```diff
@@ -79,6 +84,15 @@ module.exports = class Box {
+    ).then(data => {
+      Object.assign(this.data, data);
+    }).catch(err => {
+      this.errorStatusCode = err.statusCode;
```
Does `this.errorStatusCode` actually do anything on the page with the list of boxes? It's used to display a 404 when viewing a particular box, but it's unclear what it's doing in this context.

YaManicKill replied:
Ah, no idea. I guess not... copy-pasta

YaManicKill replied:
Removed.

YaManicKill commented:
@Raia Yeah, that's a good point, I had forgotten about the name being an indicator as to what the box is about. Good stuff.

YaManicKill commented:
Fixed the review comments.

New box card view:
![](https://i.gyazo.com/8fa7df1768479647805610df270a5c5b.png)

I'm pretty happy with this now. I made it slightly smaller so that we don't have part of a pokemon appearing at the bottom (awkward), and I've changed the padding to allow us to have a few more pokemon, and make them properly spaced.

I think it's looking pretty good now. Thoughts?

TheSonAlsoRises commented:
Looking good!

not-an-aardvark commented:
Planning to review this again tomorrow.

YaManicKill commented:
@not-an-aardvark Great, thanks. I'm going to do some more testing on it to make sure it all works fine.

not-an-aardvark commented on client/add/add.ctrl.js:
```diff
@@ -123,10 +123,27 @@ module.exports = class Add {
+        Object.keys(lines).forEach(
+          (line) => line.slice(0, BOX_PAGE_SIZE - (box.data.contents.length % BOX_PAGE_SIZE))
+        );
+        return lines;
```
I don't think this function is actually doing anything -- neither `forEach` nor `slice` has the effect of mutating `lines`. Also, the `box` variable here is the selected box, but we care about the box this particular group of Pokémon is getting uploaded to.

The goal of this function is to ensure that if a large number of Pokémon are uploaded to a box, the client just fills up the current box page rather than displaying hundreds of uploads and causing lag.

I think we want to do something like this:

```js
import {keyBy} from 'lodash';
import {mapValues} from 'lodash/fp'

const boxesById = keyBy(this.boxes, 'id');

// ...
.then(mapValues((uploadsForBox, boxId) => {
  return uploadsForBox.slice(0, BOX_PAGE_SIZE - (boxesById[boxId].data.contents.length % BOX_PAGE_SIZE))
))
```

YaManicKill replied:
Huh, yeah, you are right.

YaManicKill replied:
This becomes a lot more complex to do if we change to using a map of id to box, because this map doesn't exist anymore.

YaManicKill replied:
It's all good, I think I've found a nice, simple way to do it.

not-an-aardvark commented on client/add/add.spec.js:
```diff
@@ -9,14 +9,15 @@ Promise.config({warnings: false});
 describe('AddCtrl', () => {
   const errorHandler = (err) => err && console.log(err);
   const BOX_ID = 123;
+  const BOX_ID2 = 1234;
```
It doesn't really matter in this case, but box ids are strings.

YaManicKill replied:
Ah yeah, I'll change it just so the test is accurate. I blame whoever wrote the first test :-P

not-an-aardvark commented on client/add/add.ctrl.js:
```diff
@@ -139,11 +139,19 @@ module.exports = class Add {
         return lines;
       })
-      .then(lines => this.boxes.forEach((box) => {
+      .then(lines => {
```
If we end up creating a map of ids to boxes, it would be better to iterate over `lines` rather than all of the user's boxes.

YaManicKill replied:
I'm not sure what you are referring to. What map of ids to boxes?

YaManicKill replied:
Having thought it through, this probably makes the most sense to do. We can get the boxes by id on initial data and everytime we add a new box. That won't be much, because when we get the new box, we know the id and can just insert it. All good.

YaManicKill commented on client/add/add.ctrl.js:
```diff
@@ -123,10 +123,27 @@ module.exports = class Add {
+      .filter(line => line.success)
       .map(line => line.created)
-      .then(lines => box.data.contents.push(...lines))
+      .reduce((overall, line) => {
```
Yeah, good point. When writing the code, I couldn't figure out what to name it so I called it overall as a temp name and forgot to change it.

YaManicKill commented on client/add/add.ctrl.js:
```diff
@@ -123,10 +123,27 @@ module.exports = class Add {
+        }
+        return overall;
+      }, {})
+      .then(lines => {
```
Yeah

YaManicKill commented:
Alrighty, @not-an-aardvark, using the id->box map, it has simplified it massively, and much more efficient, I think. We just iterate over the boxes once at the beginning, add a box to the array when we add a new box, and then iterate over the inserted pokemon to insert them into the boxes.

not-an-aardvark commented on client/add/add.ctrl.js:
```diff
@@ -87,13 +89,16 @@ module.exports = class Add {
       })
-      .then(res => this.$scope.$apply(this.boxes.push(res)))
+      .tap(box => this.$scope.$apply(this.boxes.push(box)))
+      .then(box => this.$scope.$apply(this.boxesById[box.id].push(box)))
```
It looks like this was already here before, but is it intended that this will call `$scope.$apply` on a number?

YaManicKill replied:
¯\_(ツ)_/¯ I don't think it matters. It is a bit weird because we scope.$apply later on as well. I'll remove it and see if it works.

YaManicKill replied:
Ah wait, we don't do it later, it's on pokemon that we do it later.

not-an-aardvark commented on client/add/add.ctrl.js:
```diff
@@ -87,13 +89,16 @@ module.exports = class Add {
     this.watchFullScreen();
     const box = this.selected.selectedBox;
+    const boxCounts = mapValues(this.boxesById, (value) => value.contents.length);
+    const that = this;
```
I don't think `that` is necessary here because we're using arrow functions.

YaManicKill replied:
It seems to be that doing the .each does something weird with this. I'll investigate why, but it didn't work without that/.

YaManicKill replied:
Huh, alright I must have done something else wrong, it appears to work without it now. I'm muchly confused. Ach well, removed.

not-an-aardvark commented on client/add/add.ctrl.js:
```diff
@@ -123,11 +128,16 @@ module.exports = class Add {
       .map(line => line.created)
-      .then(lines => box.data.contents.push(...lines))
-      .tap(box.onscroll)
+      .filter(line => boxCounts[line.box]++ < BOX_PAGE_SIZE)
```
This isn't doing the same thing as the previous code. Previously, if a box had 45 pokemon and a bunch of new uploads were added, the amount would jump to 60 (filling the current box page). Now it will stay at 45 (because it's larger than a single box page).

I think using something along the lines of the `length % BOX_PAGE_SIZE` logic from before would be better.

(Also, I'm not really a big fan of having a filter function with side-effects, but it's not a PR-blocker.)

YaManicKill replied:
Ah right, I wasn't 100% sure about what it was meant to be doing. That makes sense.

> (Also, I'm not really a big fan of having a filter function with side-effects, but it's not a PR-blocker.)

I mean, the side-effect is purely for the filter itself, boxCounts only exists for this exact function, nothing else. So it isn't **really** a side-effect.

not-an-aardvark commented on client/add/add.ctrl.js:
```diff
@@ -123,11 +128,16 @@ module.exports = class Add {
+      .filter(line => boxCounts[line.box]++ < BOX_PAGE_SIZE)
+      .each(line => {
+        that.boxesById[line.box].contents.push(line);
+        if (box && line.box === box.id) {
```
Maybe the `box` variable would be clearer if it was renamed to `currentBox`.

YaManicKill replied:
Good plan.

YaManicKill commented:
Alright. Updated with the newest round of comments.

not-an-aardvark commented on client/add/add.ctrl.js:
```diff
@@ -123,11 +128,16 @@ module.exports = class Add {
       .map(line => line.created)
-      .then(lines => box.data.contents.push(...lines))
-      .tap(box.onscroll)
+      .filter(line => (boxCounts[line.box]++ % BOX_PAGE_SIZE) < BOX_PAGE_SIZE)
```
Now the function will always return true, because `anything % x < x`.

Just to make sure, have I explained the intended behavior clearly enough? I feel like we're miscommunicating about what this is supposed to do.

YaManicKill replied:
You have, I just mucked it up in my head.

YaManicKill replied:
I'm not sure there's a nice way to do what you want with how I'm doing the filtering. I'm also not sure of a way to do the filtering in a nice way without doing what I was doing before, which is incredibly inefficient. How wedded are you to always having it add more pokemon, even if we already have a lot?

not-an-aardvark replied:
I think it should add stuff to a box that the user is looking at. I don't care as much about whether we have a large number of icons on the homescreen.

One way to do it would be

```js
.each(line => {
  if (this.boxesById[line.box].contents.length % BOX_PAGE_SIZE) {
    this.boxesById[line.box].contents.push(line);
  }
})
```

...which is slightly inefficient because it needs to check every box rather than slicing, but I kind of doubt the difference is noticeable for a few hundred uploads.

YaManicKill replied:
Ah yes, I guess that would work, because it's adding until it gets to a multiple of hte page size and then stopping. Happy with that. I just couldn't think of a way around it, I think I had been thinking about it too much :-P

YaManicKill replied:
Ah for goodness sake, of course this won't work because 0 % x = 0. 

/me sighs

YaManicKill commented:
Alright, updated again.

not-an-aardvark commented on client/add/add.ctrl.js:
```diff
@@ -123,11 +127,18 @@ module.exports = class Add {
+        }
+        if (selectedBox && line.box === selectedBox.id
+          && selectedBox.data.contents.length % BOX_PAGE_SIZE) {
+          selectedBox.data.contents.push(line);
```
Isn't this redundant with the `if` statement above it? It looks like it will cause things to get added to the selected box twice.

YaManicKill replied:
It is not. The boxes are actually different boxes. Selectedbox is only when we have a #/box page open, and it is a different instance from the card boxes in this.boxes, which we always have (and those are what we see on the index page).

So, if we are on box 3, and we also have boxes 1 and 2, this.boxes will contain the 3 card box instances (1, 2 and 3), but selectedBox will not point to the same instance of 3 that is in this.boxes, because we create a new one when we navigate to that page.

Should we change how this works? Probably, but that's really not in the scope of this PR, as it's an existing issue. We also now wipe out selectedBox when we navigate to the homepage, so we only have a selectedBox when we are on a box page. (Also, I fixed a bug that did include duplicating pokemon because the box cards set themselves to be selectedBox, which is clearly nonsense).

YaManicKill commented:
Alrighty, I'll just fix the test that I just broke and then I'll merge. Thanks, @not-an-aardvark.

YaManicKill commented:
Alrighty, fixed the tests, and the box size ones were easy enough to do so I added them in.
